lms/djangoapps/grades/rest_api/v1/views.py

```diff
@@ -207,3 +217,307 @@ def _get_course(self, request, course_id):
+                            * state: State of submission.
+                            * grade: Grade.
+                            * max_grade: Maximum possible grade.
+                        * data: problem's data.
+        """
+        data = []
+        username = request.user.username
+        all_users = False
+        if GlobalStaff().has_user(request.user):
+            username = request.GET.get('username', request.user.username)
+            all_users = bool(request.GET.get('all_users', False))
+        try:
+            course_id = get_course_key(request, course_id)
+        except InvalidKeyError:
+            raise self.api_error(  # lint-amnesty, pylint: disable=raise-missing-from
+                status_code=status.HTTP_400_BAD_REQUEST,
+                developer_message='The provided course key cannot be parsed.',
+                error_code='invalid_course_key'
+            )
+        if (GlobalStaff().has_user(request.user) or self.has_api_key_permissions(request)):
+            course_enrollments = CourseEnrollment.objects.select_related('user').filter(
+                is_active=True
+            )
+            if course_id:
+                course_enrollments = course_enrollments.filter(
+                    course_id=course_id
+                ).order_by('created')
+
+            if not all_users:
+                course_enrollments = course_enrollments.filter(user__username=username)
+
+            course_xblock_structure = self. _generate_course_xblock_structure(course_enrollments)
+            for course_key, course_info in course_xblock_structure.items():
+                course_data = self._get_course_data(
+                    course_key,
+                    course_info.get('course_enrollment'),
+                    course_info.get('course'),
+                    course_info.get('blocks')
+                )
+                data.append(course_data)
+            return Response(data)
+        else:
+            raise self.api_error(
+                status_code=status.HTTP_403_FORBIDDEN,
+                developer_message='The course does not exist.',
+                error_code='user_or_course_does_not_exist',
+            )
+
+    def _generate_course_xblock_structure(self, course_enrollments):
+        """ Generate a map of course to course enrollment and problem
+            xblocks for each of the course.
+        """
+        course_enrollment_id_map = {}
+        course_xblock_structure = {}
+        for course_enrollment in course_enrollments:
+            course_enrollment_id_map[str(course_enrollment.course_id)] = course_enrollment
+        for course_key, course_enrollment in course_enrollment_id_map.items():
+            course = get_course(course_enrollment.course_id, depth=4)
+            blocks = self._get_problem_blocks(course)
+            course_xblock_structure[course_key] = {
+                'course_enrollment': course_enrollment,
```

Comment:
Doesn't this allow only one enrollment per course? For `all_users`, it should be a list, e.g. something like:
```python
        course_structure = {}

        for course_enrollment in course_enrollments:
            course_id = str(course_enrollment.course_id)

            if course_id not in course_structure:
                course = get_course(course_enrollment.course_id, depth=4)
                course_structure[course_id] = {
                    'enrollments': [course_enrollment],
                    'course': course, 
                    'blocks': self._get_problem_blocks(course),
                }
            else:
                course_structure[course_id]['enrollments'].append(course_enrollment)

        return course_structure
```

Reply:
There was an optimization that could be done just before that I have opted for it.